DOCS-595: Add C++ SDK (#1266)
viamrobotics/docs: docs/program/sdks/_index.md

Use the SDK of your preferred language to write code to control your robots.

Viam currently offers SDKs for the following languages:

- [Python SDK](https://python.viam.dev/)
- [Go SDK](https://pkg.go.dev/go.viam.com/rdk)
- [TypeScript SDK](https://ts.viam.dev/)

> - [C++ SDK](https://cpp.viam.dev/)

Reviewer: opt: should we note "alpha" here? Perhaps cumbersome to remember to change later, but without such notice, these SDKs seem like equal citizens in this list.

Author: ```suggestion
- [C++ SDK (alpha)](https://cpp.viam.dev/)
```

main();
```

{{% /tab %}}

> {{% tab name="C++" %}}

Reviewer: Noticed: selecting one tab on page does not auto-switch to same-name tabs everywhere on page. I can look into this!

> {{% tab name="C++" %}}

Author: ```suggestion
{{% tab name="C++" %}}

{{< alert title="Note" color="note" >}}
The C++ SDK is currently in alpha.
{{< /alert >}}
```

> ```cpp {class="line-numbers linkable-line-numbers"}

Reviewer: nit: `cpp` doesn't appear to highlight, and stands out as missing against the other three. Seems to be [the right one](https://gohugo.io/content-management/syntax-highlighting/) though (or maybe `c++` ?). I can investigate.

Reviewer: I created https://github.com/viamrobotics/docs/pull/1268 which I believe adds support for C++ highlighting. Let me know if I missed something!

Author: thank you!

# include <string>
# include <vector>

# include <boost/optional.hpp>

# include <viam/api/common/v1/common.pb.h>
# include <viam/api/robot/v1/robot.grpc.pb.h>

# include <viam/sdk/robot/client.hpp>
# include <viam/sdk/components/camera/client.hpp>

using namespace viam::sdk;

int main() {

>   std::string host("guardian-main.vw3iu72d8n.viam.cloud");

Reviewer 2: I don't think we should include an actual uri/credentials here, should we?

Author: ```suggestion
  std::string host("ADDRESS FROM THE VIAM APP");
```

  DialOptions dial_opts;

>   Credentials credentials("gf1aufai3dr0b5o9csko93b1d8llfwolisa0my141wlmvkmg");

Author: ```suggestion
  Credentials credentials("SECRET FROM THE VIAM APP");
```

- [Arm](/components/arm/#api)
- [Base](/components/base/#api)
- [Board](/components/board/#api)
- [Camera](/components/camera/#api)
- [Encoder](/components/encoder/#api)
- [Gantry](/components/gantry/#api)
- [Gripper](/components/gripper/#api)
- [Input Controller](/components/input-controller/#api)
- [Motor](/components/motor/#api)
- [Movement Sensor](/components/movement-sensor/#api)
- [Sensor](/components/sensor/#api)
- [Servo](/components/servo/#api)

{{% alert title="Tip" color="tip" %}}

> You can find example code using the SDKs in the <file>examples</file> folder of the respective repository:

Reviewer: nit: the `python` example takes us to a few subdirectories within the `examples` directory, while all other links take us directly to the `examples` directory. None of the four really require the additional step "then navigate to the `examples` folder" that this list intro seems to suggest. Perhaps just plainly stating without monospace helps to be precise for all four, something like:

> Each SDK provides example code in the following locations:

Or similar.

Author: ```suggestion
Each SDK provides example code in its repository:
```

- [Python SDK repository](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples/server/v1)
- [Go SDK repository](https://github.com/viamrobotics/rdk/tree/main/examples)
- [TypeScript SDK repository](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples)

> - [Viam C++ SDK repository](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples)

Reviewer: If you agree with above, recommend changing link titles to "X SDK examples" instead.

- [Arm](/components/arm/#api)
- [Base](/components/base/#api)
- [Board](/components/board/#api)
- [Camera](/components/camera/#api)
- [Encoder](/components/encoder/#api)
- [Gantry](/components/gantry/#api)
- [Gripper](/components/gripper/#api)
- [Input Controller](/components/input-controller/#api)
- [Motor](/components/motor/#api)
- [Movement Sensor](/components/movement-sensor/#api)
- [Sensor](/components/sensor/#api)
- [Servo](/components/servo/#api)

{{% alert title="Tip" color="tip" %}}

You can find example code using the SDKs in the <file>examples</file> folder of the respective repository:

> - [Python SDK repository](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples/server/v1)
> - [Go SDK repository](https://github.com/viamrobotics/rdk/tree/main/examples)
> - [TypeScript SDK repository](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples)
> - [Viam C++ SDK repository](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples)

Author: ```suggestion
- [Python SDK examples](https://github.com/viamrobotics/viam-python-sdk/tree/main/examples/server/v1)
- [Go SDK examples](https://github.com/viamrobotics/rdk/tree/main/examples)
- [TypeScript SDK examples](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples)
- [Viam C++ SDK examples](https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples)
```

For an example, see [this execution demo.](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples/vanilla)

{{% /tab %}}
{{% tab name="C++" %}}

> For an example, see [this execution demo.](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples/vanilla)

Reviewer: This is the TS example. I wasn't able to guess a likely C++ example from the upstream repo; not sure what this should link to.

> For an example, see [this execution demo.](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples/vanilla)

Reviewer 2: This looks like a typescript link. Probably the best link for C++ would be https://github.com/viamrobotics/viam-cpp-sdk/tree/main/src/viam/examples/dial, though we don't have a great README for it at this time.

Author: I think it actually should be https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md again. We're looking for something that points to runnign the code

For an example, see [this execution demo.](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples/vanilla)

{{% /tab %}}
{{% tab name="C++" %}}

> For an example, see [this execution demo.](https://github.com/viamrobotics/viam-typescript-sdk/tree/main/examples/vanilla)

Author: ```suggestion
For information on running C++ code see [here](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/BUILDING.md).
```
